docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,562 @@
+functionalities to it that are designed in a principled way such that:
+* tight coupling of dependencies and unrelated functionalities is minimized or eliminated
+* APIs can have long-term stability guarantees
+* the SDK framework is extensible in a safe and straightforward way
+
+The design principles of the core API are as follows:
+* everything that a module wants to interact with in the state machine is a service
+* all services coordinate state via `context.Context` and don't try to recreate the "bag of variables" approach of `sdk.Context`
+* all independent services are isolated in independent packages with minimal APIs and minimal dependencies
+* the core API should be minimalistic and designed for long-term support (LTS)
+* a "runtime" module will implement all the "core services" defined by the core API and can handle all module
+  functionalities exposed by the core `Handler` type
+* other non-core and/or non-LTS services can be exposed by specific versions of runtime modules or other modules 
+following the same design principles, this includes functionality that interacts with specific non-stable versions of
+third party dependencies such as Tendermint
+* the core API doesn't implement *any* functionality, it just defines types
+* go stable API compatibility guidelines are followed: https://go.dev/blog/module-compatibility
+
+A "runtime" module is any module which implements the core functionality of composing an ABCI app, which is currently
+handled by `BaseApp` and the `ModuleManager`. Runtime modules which implement the core API are *intentionally* separate
+from the core API in order to enable more parallel versions and forks of the runtime module than is possible with the
+SDK's current tightly coupled `BaseApp` design while still allowing for a high degree of composability and
+compatibility.
+
+Modules which are built only against the core API don't need to know anything about which version of runtime,
+`BaseApp` or Tendermint in order to be compatible. Modules from the core mainline SDK could be easily composed
+with a forked version of runtime with this pattern.
+
+This design is intended to enable matrices of compatible dependency versions. Ideally a given version of any module
+is compatible with multiple versions of the runtime module and other compatible modules. This will allow dependencies
+to be selectively updated based on battle-testing. More conservative projects may want to update some dependencies
+slower than more fast moving projects.
+
+### Core Services
+
+The following "core services" are defined by the core API. All valid runtime module implementations should provide
+implementations of these services to modules via both [dependency injection](./adr-057-app-wiring-1.md) and
+manual wiring. The individual services described below are all bundled in a convenient `appmodule.Service`
+"bundle service" so that for simplicity modules can declare a dependency on a single service.
+
+#### Store Services
```

Comment:
Could you give an example of how one would do `ctx.CacheContext()` with this store service? It doesn't seem clear to me right now

Reply:
That's maybe something that needs to be added

Reply:
I think there are two ways to do this:
1. any service which can operate in cache mode offers its own API for this, i.e. `KVStore.CacheWrap`, `EventService.CacheWrap`
2. there is some way that caching happens across multiple services which is how `CacheContext` behaves now

The downside to 1, is that you need to remember to cache all 3 store types and it isn't obvious how to do this across inter-module calls. So I think we need some way to do 2, probably with a `CacheService` that has a way of notifying other services they need to cache:

```go
type CacheService interface {
  CacheContext(context.Context) context.Context)
}
```